File: plugins/motherduck/skills/motherduck-enable-self-serve-analytics/artifacts/self_serve_rollout_example.py
import json
import sys
from pathlib import Path

import duckdb

sys.path.append(str(Path(__file__).resolve().parents[3]))

from scripts._lib.motherduck_artifact_utils import artifact_session


def fetch_rows(conn: duckdb.DuckDBPyConnection, sql: str) -> list[dict]:
    cursor = conn.execute(sql)
    columns = [col[0] for col in cursor.description]
    return [dict(zip(columns, row)) for row in cursor.fetchall()]


def main() -> None:
    with artifact_session(
        slug="motherduck-enable-self-serve-analytics",
        database_keys=["analytics"],
    ) as session:
        conn = session.conn
        accounts_table = session.table("analytics", "main", "accounts")
        customer_health_view = session.table("analytics", "main", "customer_health")
        conn.execute(
            f"""
            CREATE TABLE {accounts_table} (
                team VARCHAR,
                account_id INTEGER,
                status VARCHAR,
                arr DOUBLE
            )
            """
        )
        conn.executemany(
            f"INSERT INTO {accounts_table} VALUES (?, ?, ?, ?)",
            [
                ("sales", 1, "healthy", 12000.0),
                ("sales", 2, "watch", 7000.0),
                ("success", 3, "healthy", 9000.0),
                ("success", 4, "risk", 5000.0),
            ],
        )
        conn.execute(
            f"""
            CREATE OR REPLACE VIEW {customer_health_view} AS
            SELECT team, account_id, status, arr
            FROM {accounts_table}
            WHERE status IS NOT NULL
            """
        )

        result = {
            "backend": session.describe(),
            "first_audience": "customer success",
            "first_asset": f"team KPI Dive on top of {customer_health_view}",
            "team_kpis": fetch_rows(
                conn,
                f"""
                SELECT team, COUNT(*) AS total_accounts, SUM(arr) AS total_arr
                FROM {customer_health_view}
                GROUP BY 1
                ORDER BY total_arr DESC
                """,
            ),
        }
        print(json.dumps(result, indent=2))


if __name__ == "__main__":
    main()

File: plugins/motherduck/skills/motherduck-enable-self-serve-analytics/artifacts/self_serve_rollout_example.ts
export {};
declare const process: { env: Record<string, string | undefined> };

type AccountRow = { team: string; account_id: number; status: string; arr: number };

function normalizeMetadataValue(value: string | undefined, fallback: string): string {
  const raw = (value ?? "").trim();
  if (!raw) return fallback;
  const normalized = raw.replace(/[^A-Za-z0-9._-]+/g, "-").replace(/^[-._]+|[-._]+$/g, "");
  return normalized || fallback;
}

function buildUseCaseUserAgent(): string {
  const harness = normalizeMetadataValue(process.env.MOTHERDUCK_AGENT_HARNESS, "unknown");
  const llm = normalizeMetadataValue(process.env.MOTHERDUCK_AGENT_LLM, "unknown");
  return `agent-skills/2.2.2(harness-${harness};llm-${llm})`;
}

const accounts: AccountRow[] = [
  { team: "sales", account_id: 1, status: "healthy", arr: 12000.0 },
  { team: "sales", account_id: 2, status: "watch", arr: 7000.0 },
  { team: "success", account_id: 3, status: "healthy", arr: 9000.0 },
  { team: "success", account_id: 4, status: "risk", arr: 5000.0 },
];

const teamMap = new Map<string, { total_accounts: number; total_arr: number }>();
for (const row of accounts) {
  const current = teamMap.get(row.team) ?? { total_accounts: 0, total_arr: 0 };
  current.total_accounts += 1;
  current.total_arr += row.arr;
  teamMap.set(row.team, current);
}

const result = {
  backend: {
    mode: "typescript-companion",
    databases: { analytics: "analytics" },
    user_agent: buildUseCaseUserAgent(),
  },
  first_audience: "customer success",
  first_asset: 'team KPI Dive on top of "analytics"."main"."customer_health"',
  team_kpis: Array.from(teamMap.entries())
    .map(([team, value]) => ({ team, ...value }))
    .sort((a, b) => b.total_arr - a.total_arr),
};

console.log(JSON.stringify(result, null, 2));
